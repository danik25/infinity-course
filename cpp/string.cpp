/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}