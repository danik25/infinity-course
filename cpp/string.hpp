/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */