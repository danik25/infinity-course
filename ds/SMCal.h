#ifndef  _SMCAL_H_
#define  _SMCAL_H_


/****************************************************************
*                         calculator- state machine
*                       ==============================
* File Name: SMCal.h
* Related files: SMCal.c, SMCal_test.c
* #Version: V 1.1
* Writer: Dani Benjamin                         
* Reviewer: Ofer Tauber 
* Last update: 23/05/19
****************************************************************/

/* API Design:
 * A calculator that allows the following operators: addition, subtraction, 
 * multiplication, division, when accepting the Mathematical expression as a string.
 * If an incorrect value is inserted, the calculator throws an error.
 * the calculator is designed as state machine, that is  is a mathematical 
 * model of computation, an abstract machine that can be in exactly one of a 
 * finite number of states at any given time.
 * 
 *
 */

typedef struct calculator
{
    char message[100];
    double answer;
}cal_t;



 /* 
 *
 * A calculator that accepts numbers and operators and returns their 
 * mathematical value. If another char is entered, the calculator throws an error. 
 * The calculator is capable of dealing with spaces.
 *
 *
 */

cal_t* SMCalculator(char * exp);





#endif /* _SMCAL_H_  */