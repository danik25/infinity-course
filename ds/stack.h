/****************************************************************
*                       Data Structures - Stack                 *
* Related files: stack_test.c, stack.h, stack_test.c            * *															    *
* writer: Dani Benjamin                                         *
* reviewer: Ofer Tauber                                         *
* last update: 29.3.19                                          *
****************************************************************/        
       
#ifndef _STACK_H_
#define _STACK_H_

#include <stdio.h> /* size_t */

typedef struct stack stack_t;

/*creates the stack with capacitiy of <capacitiy>, for element of size
 <element_size>. returns a pointer to stack*/
stack_t *StackCreate(size_t element_size, size_t capacity);


/*destroy the stack and free the allocated memory*/
void StackDestroy(stack_t *stack_ptr);


/*insert element <element> to the top of the stack(by value).
*returns 0 to denote success or 1 to denote failure (pushing a full stack).*/
int StackPush(stack_t *stack_ptr, const void *element);


/*remove the element in top of tghe stack(without returning it's value)*/
void StackPop(stack_t *stack_ptr);


/*returns pointer to the top element without modifying the stack*/
void *StackPeek(const stack_t *stack_ptr);


/*returns the size of the stack without modifying it*/
size_t StackSize(const stack_t *stack_ptr);

#endif /*_STACK_H_*/
