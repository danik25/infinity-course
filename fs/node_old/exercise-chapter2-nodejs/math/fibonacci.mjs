function fibonacciSync(n)
{
    if(n <= 1){
        return 1
    }

    return (fibonacciSync(n - 1) + fibonacciSync(n - 2))
}
function callback(param, func)
{
    console.log("i: " + param + " -> " + func(param))
}

(function iifeFib()
{
    var i;
    for(i = 0; i < 5; ++i)
    {
        setImmediate(callback, i, fibonacciSync)
    }
})()

