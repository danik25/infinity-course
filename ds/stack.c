#include <stdio.h> /*printf*/
#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/
#include <string.h> /*memcpy*/

#include "stack.h"

struct stack
{
    void *end;
    void *current;
	size_t element_size;
	void *head;
};

stack_t *StackCreate(size_t element_size, size_t capacity)
{    
    stack_t *stack = malloc(sizeof(stack_t));
    stack -> head = malloc(capacity * element_size);
    
    stack -> end = (char*)stack -> head + capacity * element_size;
    stack -> element_size = element_size;
    /*current is pointing to one element size under the head*/
    stack -> current = (char*)stack -> head - element_size;
    
    if (NULL == stack || NULL == stack->head)
    {
    	return NULL;
    }
    return stack;
}

void StackDestroy(stack_t *stack_ptr)
{
	assert(stack_ptr);
	
    free(stack_ptr -> head);
    stack_ptr -> head = NULL;
    free(stack_ptr);
    stack_ptr = NULL;
}

int StackPush(stack_t *stack_ptr, const void *element)
{
	assert(stack_ptr);
	assert(element);
	/*empty stack*/
	if (stack_ptr -> current == NULL)
	{
	    stack_ptr -> current = (char*)stack_ptr -> head - stack_ptr ->
	                                element_size;
	}
	else if (stack_ptr -> head == NULL)
	{
	    fprintf(stderr, "%s", "unsuccessfull push, no stack\n\n");
	        return 1;
	}

    else if ((char*)stack_ptr -> current + stack_ptr -> element_size == 
    								stack_ptr -> end)
    {
        fprintf(stderr, "%s", "unsuccessfull push, full stack\n\n");
     	return 1; /*full stack*/
    }
    stack_ptr -> current = (char*)stack_ptr -> current + 
    								stack_ptr -> element_size;
    memcpy(stack_ptr -> current, element, stack_ptr -> element_size);
    
    return 0;
}

void StackPop(stack_t *stack_ptr)
{
	assert(stack_ptr);
	
    if (stack_ptr -> current < stack_ptr -> head)
    {
        fprintf(stderr, "%s", "unsuccessfull pop, empty stack\n\n");
    }
    else
    {
        stack_ptr -> current = (char*)stack_ptr -> current - stack_ptr ->
                                                                element_size;
        /*if it was the last element in the stack*/
        if (stack_ptr -> current < stack_ptr -> head)
        {
        	stack_ptr -> current = NULL;
        }
    }
}

void *StackPeek(const stack_t *stack_ptr)
{
	assert (stack_ptr);
    return stack_ptr->current;
}

size_t StackSize(const stack_t *stack_ptr)
{	
	size_t size = ((char*)stack_ptr->current - (char*)stack_ptr->head +
	                        stack_ptr->element_size)/stack_ptr->element_size;
	
	assert(stack_ptr);
	
	if (stack_ptr -> current < stack_ptr -> head)
	{
		return 0;
	}
	return size;
}







