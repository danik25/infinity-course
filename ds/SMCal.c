/****************************************************************
*                         calculator- state machine
*                       ==============================
* File Name: SMCal.c
* Related files: SMCal.h, SMCal_test.c
* #Version: V 1.1
* Writer: Dani Benjamin                         
* Reviewer: Ofer Tauber 
* Last update: 23/05/19
****************************************************************/

#include <assert.h> /* assert */
#include <string.h> /* strlen */
#include <ctype.h> /* isspace */ 
#include <math.h> /* pow */ 
#include "SMCal.h"
#include "stack.h"

#define STOP 4
#define UNUSED(x) (void)(x)

/*
START- begining
PN- proccess number
PO- proccess operand
WS- white space
*/

enum states {START, PN, PO, WS, STATES_SIZE};
enum events {END, OPERAND, DIGIT, SPACE, OTHER, EVENTS_SIZE};

typedef int (*action_func)(stack_t * stack_num, 
                        stack_t * stack_op,
                        cal_t * ans_handle,
                        const char input,
                        int state);

/* assisting functions */
static int IsOperation(char ch);
static int IsNumber(char ch);
static int Priority(char ch);
static double CurrentCalculation(double num1, double num2, char op);

/* events */
static int ConverToDigit(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state);
static int OperatorProc(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state);
static int Calculate(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state);
static int ErrorMessage(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state);
static int WSProc(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state);

cal_t* SMCalculator(char * exp)
{
    static cal_t ans_handle = {"", 0};
    int state = 0;
    int LUT[256] = {0};
    int i = 0;

    /* stacks */
    stack_t *stack_num = NULL;
    stack_t *stack_op = NULL;

    /*
        _|______|_______0_____|_______1_______|_______2_____|_______3_____|_______4_____  
         |      |  END('\0')  |   OPERAND     |    DIGIT    | SPACE (' ') |     OTHER
      ___|_____ |_____________|_______________|_____________|_____________|_____________
       0 |START | calculate   | ErrorMessage  |ConverToDigit| WSProc      |ErrorMessage
         |      | s->finish   |               | s->PN       |s->start     |
      ___|______|_____________|_______________|_____________|_____________|_____________
       1 | PN   | calculate   | OperatorProc  |ConverToDigit| WSProc      |ErrorMessage
         |      | s->finish   | s->PO         | s->PN       |s->WS        |
      ___|______|_____________|_______________|_____________|_____________|_____________
       2 | PO   | ErrorMessage| ErrorMessage  |ConverToDigit| WSProc      |ErrorMessage
         |      |             |               |s->PN        | s->PO       |
      ___|______|_____________|_______________|_____________|_____________|_____________
       3 | WS   | calculate   | OperatorProc  |ErrorMessage | WSProc      |ErrorMessage
         |      | s->finish   | s->PO         |             | s->WS       |

    */

    action_func array[STATES_SIZE][EVENTS_SIZE] = 
    {
        {Calculate, ErrorMessage, ConverToDigit, WSProc, ErrorMessage},
        {Calculate, OperatorProc, ConverToDigit, WSProc, ErrorMessage},
        {ErrorMessage, ErrorMessage, ConverToDigit, WSProc, ErrorMessage},
        {Calculate, OperatorProc, ErrorMessage, WSProc, ErrorMessage}
    };

    assert(exp);

    /* stacks */
    stack_num = StackCreate(sizeof(double), strlen(exp));
    stack_op = StackCreate(sizeof(char), strlen(exp));

    /* struct initialization */
    ans_handle.answer = 0;
    sprintf(ans_handle.message, "empty");

    /* creating lut */
    for (i = 0; i < 256; ++i)
    {
        /* 0 - 9 */
        if(IsNumber(i))
        {
            LUT[i] = DIGIT;
        }
        /* operators */
        else if(IsOperation((char)i))
        {
            LUT[i] = OPERAND;
        }
        /* '\0' */
        else if(i == END)
        {
            LUT[i] = END;
        }

        /* white spaces */
        else if(isspace(i))
        {
            LUT[i] = SPACE;
        }
        else
        {
            LUT[i] = OTHER;
        }
    }

    while(state != STOP)
    {
        state = (*array[state][LUT[(int)*exp]])(stack_num, 
                                           stack_op,
                                           &ans_handle,
                                           *exp,
                                           state);
        ++exp;
    }

    StackDestroy(stack_num);
    StackDestroy(stack_op);

    return(&ans_handle);
}

/************************* static functions ******************************/

static int ConverToDigit(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state)
{
    double num = 0;

    UNUSED(stack_num);
    UNUSED(stack_op);
    UNUSED(ans_handle);

    /* new digit to an exsisting number*/
    if(0 != StackSize(stack_num) && state == PN)
    {
        num = *(double*)StackPeek(stack_num);
        StackPop(stack_num);
    }

    num = num * 10 + (input - '0');
    StackPush(stack_num, &num);

    return (PN);
}

static int OperatorProc(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state)
{
    double num1 = 0;
    double num2 = 0;
    char sign = '0';

    UNUSED(stack_num);
    UNUSED(ans_handle);
    UNUSED(state);

    assert(stack_op);

    while( 0 != StackSize(stack_op) && 
                Priority(input) <= Priority(*(char*)StackPeek(stack_op))) 
                                            
    {
        num1 = *(double*)StackPeek(stack_num);
        StackPop(stack_num);
        num2 = *(double*)StackPeek(stack_num);
        StackPop(stack_num);
        sign = *(char*)StackPeek(stack_op);
        StackPop(stack_op);

        num1 = CurrentCalculation(num2, num1, sign);
        StackPush(stack_num, &num1);
    }

    StackPush(stack_op, &input); 
    return (PO); 
}

static int Calculate(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state)
{
    double num = 0;

    UNUSED(input);
    UNUSED(state);

    assert(stack_num);
    assert(stack_op);
    assert(ans_handle);

    if(state != START)
    {
        num = *(double*)StackPeek(stack_num);
        StackPop(stack_num);

        while(0 != StackSize(stack_num))
        {
            num = CurrentCalculation(*(double*)StackPeek(stack_num), num, 
                                            *(char*)StackPeek(stack_op));
            StackPop(stack_num);
            StackPop(stack_op);
        }    
    }

    ans_handle->answer = num;

    return (STOP);
}

static int ErrorMessage(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state)
{
    UNUSED(stack_num);
    UNUSED(stack_op);
    UNUSED(ans_handle);
    UNUSED(input);

    if(state == PO || state == PN)
    {
        sprintf(ans_handle->message, 
            "error. number expected, problematic entery: '%c'", input);
    }
    if (state == WS)
    {
        sprintf(ans_handle->message, 
            "error. operand or NULL expected, problematic entery: '%c'", input);
    }
    ans_handle->answer = 0;

    return (STOP);
}

static int WSProc(stack_t * stack_num, 
                stack_t * stack_op,
                cal_t * ans_handle,
                const char input,
                int state)
{
    UNUSED(stack_num);
    UNUSED(stack_op);
    UNUSED(ans_handle);
    UNUSED(input);

    switch(state)
    {
        case START :
            return(START);
        case PN :
            return(WS);
        case WS :
            return(WS);
        case PO :
            return(PO);
        default :
            return (STOP);
    }
}

/*===================assistance functions ===================*/

static int IsOperation(char ch)
{
    if(ch == '*' || ch == '/' || ch == '+' || ch == '-' || ch == '^')
    {
        return(1);
    }

    return(0);
}

static int IsNumber(char ch)
{
    if(ch >= '0' && ch <= '9')
    {
        return (1);
    }

    return(0);
}

static int Priority(char ch)
{
    switch(ch)
    {
        case '+' :
            return(0);
        case '-' :
            return(0);
        case '*' :
            return(1);
        case '/' :
            return(1);
        case '^' :
            return(2);
        default :
            return (0);
    }
}

static double CurrentCalculation(double num1, double num2, char op)
{
    switch(op)
    {
        case '+' :
            return(num1 + num2);
        case '-' :
            return(num1 - num2);
        case '*' :
            return(num1 * num2);
        case '/' :
            return(num1 / num2);
        case '^' :
            return(pow(num1, num2));
        default :
            return (0);
    }
}