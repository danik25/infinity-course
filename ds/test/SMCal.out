/home/xyz/git/dani-benjamin/ds/SMCal.out